### DISCUSSION TITLE
Cannot filter by branch name

### DISCUSSION THREAD

**Discussion by eugeneko** [gitextensions/gitextensions#9689] [Q&A]
### Environment

- Git Extensions 3.5.4.12724
- Build 65f01f39982ccb121c4df7d4b00b506939553958
- Git 2.33.1.windows.1
- Microsoft Windows NT 10.0.18363.0
- .NET Framework 4.8.4300.0
- DPI 96dpi (no scaling)

### Issue description

Note: this issue reproduces with latest releases of Git and Git Extensions.
It reproduces _only on one machine_, and it reproduces even after re-installing of both Git and Git Extensions.

The issue is that when I enter valid branch name to filter by branch, nothing happens.
The list of commits stays 100% the same.
It happens when I try to filter branch by any means, including Filter window, Branches text box, or "Show this branch only"

When I enter _invalid_ branch name, all commits disappear. Other filters are working as well.
It's only when I try to filter by actual valid branch name, nothing happens.

PS: Is it possible to 100% purge all traces of Git Extensions, so I can do true _clean_ reinstall?
I think it's some kind of tricky environment issue.

### Steps to reproduce

On my machine it reproduces directly in all cases, but I fail to reproduce it on any other machine.

### Did this work in previous version of GitExtensions?

It worked, I don't remember in which version and I don't think it's version-related.

### Diagnostics

[diag.log](https://github.com/gitextensions/gitextensions/files/7416535/diag.log)
I tried to filter by branch "sdefparm-guard".
08728165d9d1909c2acda0b2c2822ffbdd36a3d4 is actual head of this branch.

**Comment by RussKie** [comment]
This is the git command which applies the filter:
```
git -c log.showSignature=false log -z sdefparm-guard --pretty=format:"%H%T%P%n%at%n%ct%n%e%n%aN%n%aE%n%cN%n%cE%n%s%n%n%b1DEA7CC4-FB39-450A-8DDF-762FCEA28B05" --reflog --max-count=100000 --	I:\Projects\master\client\
```

Try enclosing the branch in quotes and see if that makes a difference, e.g.:
```
git -c log.showSignature=false log -z "sdefparm-guard" --pretty=format:"%H%T%P%n%at%n%ct%n%e%n%aN%n%aE%n%cN%n%cE%n%s%n%n%b1DEA7CC4-FB39-450A-8DDF-762FCEA28B05" --reflog --max-count=100000 --	I:\Projects\master\client\
```

**Comment by eugeneko** [comment]
Thanks for the hint, I poked this command in terminal and realized that this issue was caused by `--reflog`.
I wonder what's the intended behavior. I expected filter to take priority over reflog, but apparently it's the other way around.

Is it normal Git behavior or is it a Git bug?
If it's normal behavior, perhaps Git Extensions UI can somehow hint that these two things are incompatible.
Because when I click "show this branch only" and nothing happens, it totally feels like a bug.

**Comment by pmiossec** [comment]
Yes, that's the normal behavior. The reflog option is not something that should be enabled most of the time because that's something to retrieve lost commits and it displays **all the commits** available in git repostitory even the one not in branches (but that are in the reflog)

So that's normal that filtering has no effects because the reflog display ALL commits whatever you decide to display only a branch or all.

Maybe you're a little lost because you don't really know the git reflog feature and you checked it maybe by mistake.

>If it's normal behavior, perhaps Git Extensions UI can somehow hint that these two things are incompatible.

A good question is how?

So I think we can close the issue because that's the expected behavior...

**Comment by pmiossec** [comment]
> The reflog option is not something that should be enabled most of the time

To be clearer, that's something you should enable only when you want to retrieve a lost commit (not pointed by any reference -- i.e. branches, remotes, tags--)

**Comment by eugeneko** [comment]
> So I think we can close the issue because that's the expected behavior...

I agree that Git behavior is expected.

I don't agree Git Extensions behaviour is expected.
When I click button called "Show this branch only", I reasonably expect to see "this branch only".

> The reflog option is not something that should be enabled most of the time

I used reflog to restore lost commit some time ago and then I totally forgot about this option.

> A good question is how?

Yep...

"Show this branch only" should either override "reflog", or be disabled, or emit an error, because these two features clearly cannot work simultaneously.

Filter branch, both in Filter window and that textbox... I guess it can emit an error popup when it has non-empty branch when reflog is on? Or maybe override reflog as well.

I realise that it's minor rare issue, so I'm not asking anyone to actually implement this, but I _think_ that behaviours I listed above are objectively better than current behaviour.

**Comment by gerhardol** [comment]
In master, reflog is separate button, to both simplify usage and to visualize usage.
Reflog overrides, this could disable the filter controls though.

**Reply by RussKie** [reply]
![image](https://user-images.githubusercontent.com/4403806/138986154-b751b94f-4d79-4390-b502-c52384892587.png)
The "Show current branch" unsets reflogs.

**Reply by eugeneko** [reply]
In my case, the issue was caused by forgotten View/Show reflog references, and it isn't reset in latest release

**Reply by pmiossec** [reply]
@eugeneko What @RussKie and @gerhardol are telling you is that it is better handled in the currently on development version (i.e `master` branch).
So you'd better test the [dev portable version](https://github.com/gitextensions/gitextensions#downloads) if you want to check if it is working as you expect or if it need to be improved.
